mixer/pkg/adapter/instanceUtil.go

```diff
@@ -0,0 +1,202 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package adapter
+
+import (
+	"fmt"
+	"net"
+	"reflect"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"istio.io/istio/mixer/pkg/pool"
+)
+
+// This file contains common utility functions that adapters need to process content
+// of instances.
+
+// Stringify converts basic data types, supported by `istio.mixer.v1.config.descriptor.ValueType`, into string.
+// Note nil object is converted into empty "" string.
+func Stringify(v interface{}) string {
+	if v != nil {
+		// Switch on all possible ValueTypes
+		switch vv := v.(type) {
+		case string:
+			return vv
+		case int64:
+			return strconv.FormatInt(vv, 10)
+		case float64:
+			return strconv.FormatFloat(vv, 'f', -1, 64)
+		case bool:
+			return strconv.FormatBool(vv)
+		case time.Time:
+			return string(serializeTime(vv))
+		case time.Duration:
+			return vv.String()
+		case net.IP:
+			return vv.String()
+		case EmailAddress:
+			return string(vv)
+		case URI:
+			return string(vv)
+		case DNSName:
+			return string(vv)
+		case map[string]string:
+			buffer := pool.GetBuffer()
+			keys := make([]string, 0, len(vv))
+			for k := range vv {
+				keys = append(keys, k)
+			}
+			sort.Strings(keys)
+			for _, k := range keys {
+				if buffer.Len() != 0 {
+					buffer.WriteString("&")
+				}
+				buffer.WriteString(k)
+				buffer.WriteString("=")
+				buffer.WriteString(vv[k])
+			}
+			return buffer.String()
```

Comment:
Getting close :-)

You need to do a pool.PutBuffer before returning so put the buffer back into the pool.

Reply:
Dang it. fixed